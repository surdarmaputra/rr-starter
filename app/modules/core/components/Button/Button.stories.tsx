import type { Meta, StoryObj } from '@storybook/react';

import { Button, type ButtonProps } from './Button';

type Story = StoryObj<typeof Button>;

const meta: Meta = {
  component: Button,
  title: 'Core/Components/Button',
};

export default meta;

export const Default: Story = {
  args: {
    children: 'Button',
  },
  render: (args: ButtonProps) => <Button {...args} />,
};
