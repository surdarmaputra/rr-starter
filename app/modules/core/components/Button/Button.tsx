import { Slot } from '@radix-ui/react-slot';
import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';

import { cn } from '~/modules/core/libs/shadcn/utils';

import type { TestableComponentProps } from '../../types';

const buttonVariants = cva(
  `
    inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-full text-sm font-semibold
    ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2
    focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none
    disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0
    dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300
  `,
  {
    variants: {
      variant: {
        primary: 'bg-primary text-white hover:bg-primary-hover',
        'primary-outlined':
          'border border-primary text-primary hover:bg-primary-hover hover:text-white',
        'primary-ghost': 'text-primary hover:bg-primary-hover hover:text-white',
        secondary: 'bg-secondary text-white hover:bg-secondary-hover',
        'secondary-outlined':
          'border border-secondary text-secondary hover:bg-secondary-hover hover:text-white',
        'secondary-ghost':
          'text-secondary hover:bg-secondary-hover hover:text-white',
        success: 'bg-success text-white hover:bg-success-hover',
        'success-outlined':
          'border border-success text-success hover:bg-success-hover hover:text-white',
        'success-ghost': 'text-success hover:bg-success-hover hover:text-white',
        warning: 'bg-warning text-white hover:bg-warning-hover',
        'warning-outlined':
          'border border-warning text-warning hover:bg-warning-hover hover:text-white',
        'warning-ghost': 'text-warning hover:bg-warning-hover hover:text-white',
        danger: 'bg-danger text-white hover:bg-danger-hover',
        'danger-outlined':
          'border border-danger text-danger hover:bg-danger-hover hover:text-white',
        'danger-ghost': 'text-danger hover:bg-danger-hover hover:text-white',
        link: 'text-textTitle underline-offset-4 hover:underline',
      },
      size: {
        default: 'h-10 px-4 py-2',
        sm: 'h-9 px-3',
        lg: 'h-11 px-8',
        icon: 'h-10 w-10',
      },
    },
    defaultVariants: {
      variant: 'primary',
      size: 'default',
    },
  },
);

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants>,
    TestableComponentProps {
  asChild?: boolean;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, variant, size, asChild = false, ...props }, ref) => {
    const Comp = asChild ? Slot : 'button';
    return (
      <Comp
        className={cn(buttonVariants({ variant, size, className }))}
        ref={ref}
        {...props}
      />
    );
  },
);

Button.displayName = 'Button';

export { Button, buttonVariants };
