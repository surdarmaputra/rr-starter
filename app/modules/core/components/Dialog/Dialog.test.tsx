import { render, screen, within } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import { Button } from '../Button/Button';
import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogFooter,
  DialogHeader,
  type DialogProps,
  DialogTitle,
} from './Dialog';

describe('Dialog', () => {
  function renderComponent(props?: DialogProps) {
    return render(
      <Dialog {...props}>
        <DialogContent data-testid="DialogContent-testid">
          <DialogHeader data-testid="DialogHeader-testid">
            <DialogTitle>Are you absolutely sure?</DialogTitle>
            <DialogDescription>Header description</DialogDescription>
          </DialogHeader>
          <div data-testid="DialogContentBody-testid">Content body</div>
          <DialogFooter data-testid="DialogFooter-testid">
            <Button type="submit">Save changes</Button>
          </DialogFooter>
        </DialogContent>
      </Dialog>,
    );
  }

  test('hides content when not opened', async () => {
    renderComponent();
    expect(
      screen.queryByTestId('DialogContent-testid'),
    ).not.toBeInTheDocument();
  });

  test('displays content when opened', async () => {
    renderComponent({
      open: true,
    });
    const dialog = await screen.findByTestId('DialogContent-testid');

    const header = within(dialog).getByTestId('DialogHeader-testid');
    within(header).getByText(/Are you absolutely sure?/);
    within(header).getByText(/Header description/);

    const body = within(dialog).getByTestId('DialogContentBody-testid');
    within(body).getByText(/Content body/);

    const footer = within(dialog).getByTestId('DialogFooter-testid');
    within(footer).getByRole('button', { name: /Save changes/ });
  });

  test('renders snapshot', async () => {
    renderComponent({
      open: true,
    });
    const dialog = await screen.findByTestId('DialogContent-testid');
    expect(dialog).toMatchSnapshot();
  });
});
