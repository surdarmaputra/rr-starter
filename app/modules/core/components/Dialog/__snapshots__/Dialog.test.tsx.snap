// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dialog > renders snapshot 1`] = `
<div
  aria-describedby="radix-:r8:"
  aria-labelledby="radix-:r7:"
  class="fixed left-[50%] top-[50%] z-50 grid w-full max-w-lg translate-x-[-50%] translate-y-[-50%] gap-4 border border-slate-200 bg-white p-6 shadow-lg duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%] data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%] dark:border-slate-800 dark:bg-slate-950 sm:rounded-lg"
  data-state="open"
  data-testid="DialogContent-testid"
  id="radix-:r6:"
  role="dialog"
  style="pointer-events: auto;"
  tabindex="-1"
>
  <div
    class="flex flex-col space-y-1.5 text-center sm:text-left"
    data-testid="DialogHeader-testid"
  >
    <h2
      class="text-lg font-semibold leading-none tracking-tight"
      id="radix-:r7:"
    >
      Are you absolutely sure?
    </h2>
    <p
      class="text-sm text-slate-500 dark:text-slate-400"
      id="radix-:r8:"
    >
      Header description
    </p>
  </div>
  <div
    data-testid="DialogContentBody-testid"
  >
    Content body
  </div>
  <div
    class="flex flex-col-reverse sm:flex-row sm:justify-end sm:space-x-2"
    data-testid="DialogFooter-testid"
  >
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-full text-sm font-semibold ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 bg-primary text-white hover:bg-primary-hover h-11 px-4 py-2"
      type="submit"
    >
      Save changes
    </button>
  </div>
  <button
    class="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-white transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-slate-950 focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-slate-100 data-[state=open]:text-slate-500 dark:ring-offset-slate-950 dark:focus:ring-slate-300 dark:data-[state=open]:bg-slate-800 dark:data-[state=open]:text-slate-400"
    type="button"
  >
    <svg
      class="lucide lucide-x h-4 w-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M18 6 6 18"
      />
      <path
        d="m6 6 12 12"
      />
    </svg>
    <span
      class="sr-only"
    >
      Close
    </span>
  </button>
</div>
`;
